glean-core/src/scheduler.rs

```diff
@@ -0,0 +1,314 @@
+//!
+//! The Metrics Ping Scheduler (MPS) is responsible for scheduling "metrics" pings.
+//! It implements the spec described in
+//! [the docs](https://mozilla.github.io/glean/book/user/pings/metrics.html#scheduling)
+
+use crate::metrics::{DatetimeMetric, StringMetric, TimeUnit};
+use crate::{CommonMetricData, Glean, Lifetime, INTERNAL_STORAGE};
+use chrono::prelude::*;
+use chrono::Duration;
+use once_cell::sync::OnceCell;
+use std::sync::{
+    atomic::{AtomicU32, Ordering},
+    Arc,
+};
+
+trait Policy {
+    fn app_build(&self, glean: &Glean) -> String;
+
+    fn get_last_sent_build(&self, glean: &Glean) -> Option<String>;
+    fn set_last_sent_build(&self, glean: &Glean, value: &str);
+    fn get_last_sent_time(&self, glean: &Glean) -> Option<DateTime<FixedOffset>>;
+    fn set_last_sent_time(&self, glean: &Glean, value: DateTime<Local>);
+
+    fn schedule_on_thread(
+        &self,
+        sleep_duration: Duration,
+        generation: u32,
+        reason: Option<&'static str>,
+    );
+}
+
+impl<'a, T> Policy for &'a T
+where
+    T: Policy,
+{
+    fn app_build(&self, glean: &Glean) -> String {
+        (*self).app_build(&glean)
+    }
+
+    fn get_last_sent_build(&self, glean: &Glean) -> Option<String> {
+        (*self).get_last_sent_build(&glean)
+    }
+    fn set_last_sent_build(&self, glean: &Glean, value: &str) {
+        (*self).set_last_sent_build(&glean, value)
+    }
+    fn get_last_sent_time(&self, glean: &Glean) -> Option<DateTime<FixedOffset>> {
+        (*self).get_last_sent_time(&glean)
+    }
+    fn set_last_sent_time(&self, glean: &Glean, value: DateTime<Local>) {
+        (*self).set_last_sent_time(&glean, value)
+    }
+
+    fn schedule_on_thread(
+        &self,
+        sleep_duration: Duration,
+        generation: u32,
+        reason: Option<&'static str>,
+    ) {
+        (*self).schedule_on_thread(sleep_duration, generation, reason)
+    }
+}
+
+struct GleanPolicy {}
+
+impl Policy for GleanPolicy {
+    fn app_build(&self, glean: &Glean) -> String {
+        glean.app_build.as_ref().unwrap().to_string()
+    }
+
+    fn get_last_sent_build(&self, glean: &Glean) -> Option<String> {
+        get_last_sent_build_metric().get_value(&glean, INTERNAL_STORAGE)
+    }
+    fn set_last_sent_build(&self, glean: &Glean, value: &str) {
+        get_last_sent_build_metric().set(&glean, value)
+    }
+    fn get_last_sent_time(&self, glean: &Glean) -> Option<DateTime<FixedOffset>> {
+        get_last_sent_time_metric().get_value(&glean, INTERNAL_STORAGE)
+    }
+    fn set_last_sent_time(&self, glean: &Glean, value: DateTime<Local>) {
+        get_last_sent_time_metric().set(&glean, Some(value.into()))
+    }
+
+    fn schedule_on_thread(
+        &self,
+        sleep_duration: Duration,
+        generation: u32,
+        reason: Option<&'static str>,
+    ) {
+        let thread_gen = Arc::clone(&THREAD_GENERATION.get_or_init(Default::default));
+        std::thread::Builder::new()
```

Comment:
So I'm a bit worried about this new thread. It's long running and sleeping all the time.
I _think_ we had the problem before that Firefox tooling wants us to ensure threads are stopped before the main thread. As it stands we don't have a way to inform this thread.

Reply:
(Additionally I'd wish we would just have a way to hook into whatever-existing background-threaded timers the surrounding thing gives us :( )